src/proxy.js

```diff
@@ -282,9 +377,19 @@ function peek(draft, prop) {
 }
 
 function markChanged(state) {
+    let makeCopyFn = assign
+    if (isMap(state.base)) {
+        makeCopyFn = assignMap
+    } else if (isSet(state.base)) {
+        makeCopyFn = assignSet
+    }
     if (!state.modified) {
         state.modified = true
-        state.copy = assign(shallowCopy(state.base), state.drafts)
+        state.copy = makeCopyFn(shallowCopy(state.base), state.drafts)
+        // state.copies is present for Sets only
+        if (state.copies !== undefined) {
+            state.copies = state.drafts
```

Comment:
If we can, we should use the `drafts` property after `modified = true` (for `Set` only) instead of adding a new `copies` property. Is there anything preventing us from doing that?

Reply:
`drafts` is being set to `null`. I figured an additional property for Sets would be easier to understand than a conditional assignment of `drafts`  to null. Though I don't have any strict preference.

Reply:
I think the existence of both `drafts` and `copies` is confusing, since `copies` may contain drafts.